token-client-core/src/main/java/no/nav/security/token/support/client/core/oauth2/AbstractOAuth2TokenClient.java

```diff
@@ -76,16 +75,29 @@ OAuth2AccessTokenResponse getTokenResponse(T grantRequest) {
         Map<String, String> formParameters = new LinkedHashMap<>();
         ClientAuthenticationProperties auth = clientProperties.getAuthentication();
         if (ClientAuthenticationMethod.CLIENT_SECRET_POST.equals(auth.getClientAuthMethod())) {
-
             formParameters.put(OAuth2ParameterNames.CLIENT_ID, auth.getClientId());
             formParameters.put(OAuth2ParameterNames.CLIENT_SECRET, auth.getClientSecret());
 
         } else if (ClientAuthenticationMethod.PRIVATE_KEY_JWT.equals(auth.getClientAuthMethod())) {
-
             ClientAssertion clientAssertion = new ClientAssertion(clientProperties.getTokenEndpointUrl(), auth);
-            formParameters.put(OAuth2ParameterNames.CLIENT_ID, auth.getClientId());
-            formParameters.put(OAuth2ParameterNames.CLIENT_ASSERTION_TYPE, clientAssertion.assertionType());
-            formParameters.put(OAuth2ParameterNames.CLIENT_ASSERTION, clientAssertion.assertion());
+
+            if (clientProperties.getGrantType().equals(OAuth2GrantType.TOKEN_EXCHANGE)) {
+                ExchangeProperties exchangeProperties = clientProperties.getTokenExchange();
+                formParameters.put(OAuth2ParameterNames.CLIENT_ID, auth.getClientId());
+                formParameters.put(OAuth2ParameterNames.CLIENT_ASSERTION_TYPE, clientAssertion.assertionType());
+                formParameters.put(OAuth2ParameterNames.CLIENT_ASSERTION, clientAssertion.assertion());
+                formParameters.put(OAuth2ParameterNames.SUBJECT_TOKEN_TYPE, exchangeProperties.subjectTokenType());
+                formParameters.put(OAuth2ParameterNames.SUBJECT_TOKEN, exchangeProperties.getSubjectToken());
+                formParameters.put(OAuth2ParameterNames.AUDIENCE, exchangeProperties.getAudience());
+                if (exchangeProperties.getResource() != null && !exchangeProperties.getResource().isEmpty()) {
+                    formParameters.put(OAuth2ParameterNames.RESOURCE, exchangeProperties.getResource());
+                }
+
+            } else {
+                formParameters.put(OAuth2ParameterNames.CLIENT_ID, auth.getClientId());
+                formParameters.put(OAuth2ParameterNames.CLIENT_ASSERTION_TYPE, clientAssertion.assertionType());
+                formParameters.put(OAuth2ParameterNames.CLIENT_ASSERTION, clientAssertion.assertion());
+            }
```

Comment:
Could drop the `else` here and extract these out as they're the same in both paths.

Reply:
yeah. agree